Resource management- Refactored

Sharvaribarve opened the pull request:
This branch contains the refactored code for ext-properties, hotdeploy and error in resources

Sharvaribarve committed:
Added External Properties test

Sharvaribarve committed:
Added general methods for resources and external properties

Sharvaribarve committed:
Reverted upload resource file

Sharvaribarve committed:
Changed naming

Sharvaribarve committed:
Added documentation and added a parameter to function

Sharvaribarve committed:
Added function to click an element when its clickable

Sharvaribarve committed:
Added resource error handling and common run steps for editing resource box

Sharvaribarve committed:
Added scenarios for resourceError Management and added files for hotDeploy

Sharvaribarve committed:
Changed feature file name

Sharvaribarve committed:
Added hotDeploy functions and scenarios

Sharvaribarve committed:
Removed unwanted methods

Sharvaribarve committed:
Added jvm hotDeploy and table to cleanup.sql

Sharvaribarve committed:
Changed documentation and removed lines

Sharvaribarve committed:
Refactored scenarios and added methods

Sharvaribarve committed:
Added common functions and added missing parts to scenarios

Sharvaribarve committed:
Removed unwanted part of the scenario

Sharvaribarve committed:
Fixed typo

Sharvaribarve committed:
Removed unwanted scenarios and refactored steps

Sharvaribarve committed:
Combined scenarios and added common functions

Sharvaribarve committed:
Added common functions and refactored scenarios accordingly

Sharvaribarve committed:
Reformatted code and Removed extra lines

Sharvaribarve committed:
Combined scenarios

Sharvaribarve committed:
Changed scenario

jeddcuison commented on jwala-integration-test/src/test/resources/com/cerner/jwala/ui/selenium/configuration/resourceErrorHandling.feature:
```diff
@@ -206,7 +206,7 @@ Feature: deploying a resource template with error produces the appropriate error
     Then I confirm multiple resource deploy error popup for file "server.xml" and file "setenv.bat" and jvm "seleniumJvm"
 
-  Scenario: an resource with a garbage value in group jvm resource
+      Scenario: an resource with a garbage value in group jvm resource
```
Should be "A resource" instead of "An resource" since 'r' is a consonant. As is it is also very hard to understand what the scenario is trying to say. I read the steps and from the looks of it you are trying deploy a resource template with invalid characters. If that's the case, it is easier to understand if you just say "Deploy resource templates with invalid characters".

jeddcuison commented on jwala-integration-test/src/test/java/com/cerner/jwala/ui/selenium/steps/configuration/ControlResourcesRunSteps.java:
```diff
@@ -0,0 +1,7 @@
+/**
+ * Created by SB053052 on 9/2/2017.
+ */
+public class ControlResourcesRunSteps {
```
Next time please don't commit the class if it is empty or non functional. It would only appear as clutter in history.

jeddcuison commented on jwala-integration-test/src/test/java/com/cerner/jwala/ui/selenium/configuration/ExternalPropertiesTest.java:
```diff
@@ -0,0 +1,15 @@
+@RunWith(Cucumber.class)
+@CucumberOptions(
+        features = {"classpath:com/cerner/jwala/ui/selenium/configuration/externalProperties.feature"},
+        glue = {"com.cerner.jwala.ui.selenium.steps"})
```
Please add the following:

    strict = true

jeddcuison commented on jwala-integration-test/src/test/java/com/cerner/jwala/ui/selenium/configuration/HotDeployTest.java:
```diff
@@ -0,0 +1,15 @@
+@RunWith(Cucumber.class)
+@CucumberOptions(
+        features = {"classpath:com/cerner/jwala/ui/selenium/configuration/hotDeploy.feature"},
+        glue = {"com.cerner.jwala.ui.selenium.steps"})
```
Please add the following:

    strict = true

jeddcuison commented on jwala-integration-test/src/test/java/com/cerner/jwala/ui/selenium/component/JwalaUi.java:
```diff
@@ -213,6 +213,10 @@ public void clickComponentForUpload(String component) {
         webElement.click();
     }
 
+    public void waitUntilElementIsClickable(final By by, final long timeout) {
```
So basically this is the same as clickWhenReady except that it does not check for the busy screen and it accepts a timeout value. I suggest that we just "overload" clickWhenReady and write a java doc to describe to users the differences of the overloaded methods. In that way the code is easier to maintain since methods with the same "general" functionality have the same name. 

Example:
```java
   /**
    * Java Doc here
    */
    public void clickWhenReady(final By by) { ... }

   /**
    * Java Doc here
    */
    public void clickWhenReady(final By by, final long timeout) { ... }
```

jeddcuison reviewed (commented):
Just a general note, I see that you named your branch with a hyphen. When you do that it is preferable that all the letters are lowercase. The said naming convention came from Unix hence the use of lower case for consistency.

With regards to commit comments, please avoid the use of "past tense". 

> The preference for present-tense, imperative-style commit messages comes from Git itself. From Documentation/SubmittingPatches in the Git repo: the body should provide a meaningful commit message, which: - uses the imperative, present tense: "change", not "changed" or "changes"

jeddcuison commented on jwala-integration-test/src/test/java/com/cerner/jwala/ui/selenium/steps/CommonRunSteps.java:
```diff
@@ -66,6 +69,11 @@
+    @Autowired
+    private ResourceErrorHandlingRunSteps resourceErrorHandlingRunSteps;
+
     private JwalaUi jwalaUi;
```
Did you run the entire tests after you're done with yours ? If you did, you'll see that clickStatusTooltipIsNotVisible breaks because jwalaUi is missing @Autowired. It is possible that jwalaUi did not come with @Autowired the time you did a rebase but even so as good developers we should been made aware of it before finalizing the branch by running the entire tests. Please rebase to master to fix this.

jeddcuison commented on jwala-integration-test/src/test/resources/com/cerner/jwala/ui/selenium/configuration/resourceErrorHandling.feature:
```diff
@@ -0,0 +1,267 @@
+
+
+    #previous metaData garbage value is erased
+    And I enter text in resource edit box and save with the following parameters:
```
I find the statement ambiguous. By resource edit box did you mean the resource text area ? Or better yet the "resource editor" ? It makes more sense to me if the step is rephrased to:

```cucumber
    And I type "${{" in the template resource editor of "APACHE-HTTPD" web server's "httpd.conf"
```
Please note that the step above is by no means final by you get my drift right ?

jeddcuison commented on jwala-integration-test/src/test/resources/com/cerner/jwala/ui/selenium/configuration/resourceErrorHandling.feature:
```diff
@@ -0,0 +1,267 @@
+      | fileName | httpd.conf |
+      | tabLabel | Template   |
+      | text     | ${{        |
+      | position | #          |
```
This defeats the purpose of Cucumber, to make things more readable to less technical people and to those who did not write the test. The relationship of the "position" key and the value "#" is not obvious to me or to the casual observer unless we dig in the code.

Sharvaribarve committed:
Added overloaded function, added strict condition and changed wording

Sharvaribarve committed:
Added strict condition to externalPropertiesTest

jeddcuison commented on jwala-integration-test/src/test/java/com/cerner/jwala/ui/selenium/steps/CommonRunSteps.java:
```diff
@@ -233,17 +241,190 @@ public void generateWebAppOfGroup(final String webAppName, final String groupNam
+
+    /*
+      The methods below don't have the check for success, so we can use them both for error checking and succesfuly deploy
+     */
```
Just a minor note. I normally let IntelliJ do the java doc comment by typing /** then enter. Your comment will then look like this:

    /**
     *  Your comment here. This would look like the standard Java Doc comment.
     */

jeddcuison commented on jwala-integration-test/src/test/resources/com/cerner/jwala/ui/selenium/configuration/externalProperties.feature:
```diff
@@ -0,0 +1,58 @@
+    And I click the ok button
+    Then I check for resource "ext.properties"
+    When I expand "ext" node in data tree
+    Then I verify dataTree has the key-value pair as "request.queue.jms.listener.receive.timeout" and "60000" respectively
```
Minor, I'd leave our "respectively" since it's very easy to understand already

jeddcuison commented on jwala-integration-test/src/test/resources/com/cerner/jwala/ui/selenium/configuration/externalProperties.feature:
```diff
@@ -0,0 +1,58 @@
+    And I click the add resource button
+    And I choose the resource file "hctProperties.tpl"
+    And I click the ok button
+    When I click "Ext Properties" component
```
If there's no "Then" here, the "When" here can be an "And" to show continuity of a workflow even if that particular step has @When

jeddcuison commented on jwala-integration-test/src/test/resources/com/cerner/jwala/ui/selenium/configuration/externalProperties.feature:
```diff
@@ -0,0 +1,58 @@
+    And I select the resource file "ext.properties"
+    And I add property "team=ctp"
+    And I click save button of edit box of "Template"
+    And I wait for notification "Saved"
```
Should this be "Then" ?

jeddcuison commented on jwala-integration-test/src/test/resources/com/cerner/jwala/ui/selenium/configuration/handleResource.feature:
```diff
@@ -0,0 +1,102 @@
+Feature: Deleting a resource
+
+  Scenario: Delete a resource -individual jvm
```
Minor, I'd just say here "Delete a JVM's resource" and for the JVM section I'd say "Delete a JVM level resource"

jeddcuison committed:
Remove unused import

jeddcuison committed:
Rename the resource files

This will make them consistent with the existing ones which uses hyphens to separate words and period to separate filename extensions

jeddcuison committed:
Correct the spelling of "succesful"

 In addition update the comments to make them more readable

jeddcuison committed:
Fix failing test

Fails in the insert of invalid characters in set env bat template since it's not able to find the "position" keyword SET CATALINA_HOME=${jvm.tomcatMedia.remoteDir}/${jvm.jvmName}/${jvm.tomcatMedia.rootDir}

jeddcuison reviewed (approved):
Minor complaint is that in resourceErrorHandling.feature, when inserting invalid characters, the position column pertains to a keyword in the template which is confusing at best when somebody reads it
